### ISSUE TITLE
goto &subroutine leaks memory

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6288]
Migrated from [rt.perl.org#20777](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=20777) (status was 'resolved')

Searchable as RT20777$

**Comment by p5pRT** [comment]
### From christian@pflanze.mine.nu

This is under 5\.6\.1 \(debian woody\, I don't have 5\.8 at hand\)\.

Cheers
Christian\.

\(Why I've run into this? I've tried to play with ideas outlined at 
http&#8203;://www\.ps\.uni\-sb\.de/~duchier/python/continuations\.html \)


\#\!/usr/bin/perl \-w
\# Fre Feb  7 11&#8203;:40&#8203;:48 MET 2003
\# christian\.jaeger@&#8203;ethlife\.ethz\.ch

sub A \{
&nbsp;     \#print "A&#8203;: $\_\[0\]\\n";
&nbsp;     $\_\[0\]\+\+;
&nbsp;     goto &B;
\}
sub B \{
&nbsp;     \#print "B&#8203;: $\_\[0\]\\n";
&nbsp;     if \($\_\[0\]\+\+ > 1000000\)\{
&nbsp;    return;
&nbsp;     \}
&nbsp;     goto &A;
\}

use BSD&#8203;::Resource;
setrlimit\(RLIMIT\_VMEM\,7\_000\_000\,7\_000\_000\) or die "could not set ulimit&#8203;: $\!";
sub xfork\{
&nbsp;     defined \(my $chld=fork\) or die $\!;
&nbsp;     $chld
\}
\# check if limit works&#8203;:
if \(xfork\)\{
&nbsp;     wait;
&nbsp;     if \($?==0\)\{
&nbsp;    die "resource limit does not work";
&nbsp;     \}
\} else \{
&nbsp;     my @&#8203;a;
&nbsp;     for\(1\.\.100\)\{
&nbsp;    push @&#8203;a\,"x"x100000;
&nbsp;     \}
&nbsp;     exit 0;
\}

\# the test&#8203;:
if \(xfork\)\{
&nbsp;     wait;
&nbsp;     $?==0 ? print "ok\\n" : print "not ok\\n";
\} else \{
&nbsp;     my $counter=1;
&nbsp;     A\($counter\);
&nbsp;     \#print "finished\.\\n";
&nbsp;     \#sleep 10000;
&nbsp;     exit 0;
\}

\# end test\.

**Comment by p5pRT** [comment]
### From @eserte

Christian Jaeger \(via RT\) \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:

> \# New Ticket Created by  Christian Jaeger 
> \# Please include the string&#8203;:  \[perl \#20777\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt2/Ticket/Display\.html?id=20777 >
> 
> 
> This is under 5\.6\.1 \(debian woody\, I don't have 5\.8 at hand\)\.
> 

Same problem under 5\.8\.0 and bleadperl\. Here's a shorter test case&#8203;:

A\(\);
sub A \{
&nbsp;    goto &A;
\}

Just call with

&nbsp;    limits \-d 4M perl script

I suspect it's the mortal reference created in S\_refgen but probably
never freed\.

Regards\,
&nbsp;    Slaven

> Cheers
> Christian\.
> 
> \(Why I've run into this? I've tried to play with ideas outlined at 
> http&#8203;://www\.ps\.uni\-sb\.de/~duchier/python/continuations\.html \)
> 
> 
> \#\!/usr/bin/perl \-w
> \# Fre Feb  7 11&#8203;:40&#8203;:48 MET 2003
> \# christian\.jaeger@&#8203;ethlife\.ethz\.ch
> 
> sub A \{
>      \#print "A&#8203;: $\_\[0\]\\n";
>      $\_\[0\]\+\+;
>      goto &B;
> \}
> sub B \{
>      \#print "B&#8203;: $\_\[0\]\\n";
>      if \($\_\[0\]\+\+ > 1000000\)\{
>     return;
>      \}
>      goto &A;
> \}
> 
> use BSD&#8203;::Resource;
> setrlimit\(RLIMIT\_VMEM\,7\_000\_000\,7\_000\_000\) or die "could not set ulimit&#8203;: $\!";
> sub xfork\{
>      defined \(my $chld=fork\) or die $\!;
>      $chld
> \}
> \# check if limit works&#8203;:
> if \(xfork\)\{
>      wait;
>      if \($?==0\)\{
>     die "resource limit does not work";
>      \}
> \} else \{
>      my @&#8203;a;
>      for\(1\.\.100\)\{
>     push @&#8203;a\,"x"x100000;
>      \}
>      exit 0;
> \}
> 
> \# the test&#8203;:
> if \(xfork\)\{
>      wait;
>      $?==0 ? print "ok\\n" : print "not ok\\n";
> \} else \{
>      my $counter=1;
>      A\($counter\);
>      \#print "finished\.\\n";
>      \#sleep 10000;
>      exit 0;
> \}
> 
> \# end test\.
> 
> 

\-\- 
Slaven Rezic \- slaven@&#8203;rezic\.de

&nbsp;    tkruler \- Perl/Tk program for measuring screen distances
&nbsp;    http&#8203;://ptktools\.sourceforge\.net/\#tkruler

**Comment by p5pRT** [comment]
### From enache@rdslink.ro

On Fri\, Feb 07\, 2003 at 10&#8203;:52&#8203;:40PM \+0100\, Slaven Rezic wrote&#8203;:
> Christian Jaeger \(via RT\) \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:
> 
> > \# New Ticket Created by  Christian Jaeger 
> > \# Please include the string&#8203;:  \[perl \#20777\]
> > \# in the subject line of all future correspondence about this issue\. 
> > \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt2/Ticket/Display\.html?id=20777 >
> > 
> > 
> > This is under 5\.6\.1 \(debian woody\, I don't have 5\.8 at hand\)\.
> > 
> 
> Same problem under 5\.8\.0 and bleadperl\. Here's a shorter test case&#8203;:
> 
> A\(\);
> sub A \{
>     goto &A;
> \}
> 
> Just call with
> 
>     limits \-d 4M perl script
> 
> I suspect it's the mortal reference created in S\_refgen but probably
> never freed\.

That's right\. \(pp\_refgen \-> S\_refto in fact\)\.

This seems to fix it\. Please try\.


---
<details><summary>Inline Patch</summary>

```diff
--- /arc/perl-current/pp_ctl.c	2003-02-02 19:59:19.000000000 +0200
+++ pp_ctl.c	2003-02-09 21:15:48.000000000 +0200
@@ -2102,6 +2102,7 @@ PP(pp_goto)
 	    }
 
 	    /* First do some returnish stuff. */
+	    FREETMPS;
 	    cxix = dopoptosub(cxstack_ix);
 	    if (cxix < 0)
 		DIE(aTHX_ "Can't goto subroutine outside a subroutine");
-------------------------------------------------------------------------

Regards
Adi
```
</details>

**Comment by p5pRT** [comment]
### From @eserte

Enache Adrian \<enache@&#8203;rdslink\.ro> writes&#8203;:

> On Fri\, Feb 07\, 2003 at 10&#8203;:52&#8203;:40PM \+0100\, Slaven Rezic wrote&#8203;:
> > Christian Jaeger \(via RT\) \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:
> > 
> > > \# New Ticket Created by  Christian Jaeger 
> > > \# Please include the string&#8203;:  \[perl \#20777\]
> > > \# in the subject line of all future correspondence about this issue\. 
> > > \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt2/Ticket/Display\.html?id=20777 >
> > > 
> > > 
> > > This is under 5\.6\.1 \(debian woody\, I don't have 5\.8 at hand\)\.
> > > 
> > 
> > Same problem under 5\.8\.0 and bleadperl\. Here's a shorter test case&#8203;:
> > 
> > A\(\);
> > sub A \{
> >     goto &A;
> > \}
> > 
> > Just call with
> > 
> >     limits \-d 4M perl script
> > 
> > I suspect it's the mortal reference created in S\_refgen but probably
> > never freed\.
> 
> That's right\. \(pp\_refgen \-> S\_refto in fact\)\.
> 
> This seems to fix it\. Please try\.
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\- /arc/perl\-current/pp\_ctl\.c    2003\-02\-02 19&#8203;:59&#8203;:19\.000000000 \+0200
> \+\+\+ pp\_ctl\.c    2003\-02\-09 21&#8203;:15&#8203;:48\.000000000 \+0200
> @&#8203;@&#8203; \-2102\,6 \+2102\,7 @&#8203;@&#8203; PP\(pp\_goto\)
>          \}
>  
>          /\* First do some returnish stuff\. \*/
> \+        FREETMPS;
>          cxix = dopoptosub\(cxstack\_ix\);
>          if \(cxix \< 0\)
>          DIE\(aTHX\_ "Can't goto subroutine outside a subroutine"\);
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> 

Yes\, this looks good\.

Regards\,
&nbsp;    Slaven

\-\- 
Slaven Rezic \- slaven@&#8203;rezic\.de
&nbsp;  BBBike \- route planner for cyclists in Berlin
&nbsp;  WWW version&#8203;:                           http&#8203;://www\.bbbike\.de
&nbsp;  Perl/Tk version for Unix and Windows&#8203;:  http&#8203;://bbbike\.sourceforge\.net

**Comment by p5pRT** [comment]
### From sky@nanisky.com


On söndag\, feb 9\, 2003\, at 21&#8203;:10 Europe/Stockholm\, Enache Adrian wrote&#8203;:

> On Fri\, Feb 07\, 2003 at 10&#8203;:52&#8203;:40PM \+0100\, Slaven Rezic wrote&#8203;:
>> Christian Jaeger \(via RT\) \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:
>>
>>> \# New Ticket Created by  Christian Jaeger
>>> \# Please include the string&#8203;:  \[perl \#20777\]
>>> \# in the subject line of all future correspondence about this issue\.
>>> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt2/Ticket/Display\.html?id=20777 >
>>>
>>>
>>> This is under 5\.6\.1 \(debian woody\, I don't have 5\.8 at hand\)\.
>>>
>>
>> Same problem under 5\.8\.0 and bleadperl\. Here's a shorter test case&#8203;:
>>
>> A\(\);
>> sub A \{
>>     goto &A;
>> \}
>>
>> Just call with
>>
>>     limits \-d 4M perl script
>>
>> I suspect it's the mortal reference created in S\_refgen but probably
>> never freed\.
>
> That's right\. \(pp\_refgen \-> S\_refto in fact\)\.
>
> This seems to fix it\. Please try\.
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\- 
> \-
> \-\-\- /arc/perl\-current/pp\_ctl\.c    2003\-02\-02 19&#8203;:59&#8203;:19\.000000000 \+0200
> \+\+\+ pp\_ctl\.c    2003\-02\-09 21&#8203;:15&#8203;:48\.000000000 \+0200
> @&#8203;@&#8203; \-2102\,6 \+2102\,7 @&#8203;@&#8203; PP\(pp\_goto\)
>          \}
>
>          /\* First do some returnish stuff\. \*/
> \+        FREETMPS;
>          cxix = dopoptosub\(cxstack\_ix\);
>          if \(cxix \< 0\)
>          DIE\(aTHX\_ "Can't goto subroutine outside a subroutine"\);
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\- 
> \-\-
>
> Regards
> Adi
>

Thanks\, applied as 18688

Arthur

**Comment by p5pRT** [comment]
arthur@contiller.se - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From arthur@contiller.se

Hi\, this bug is fixed in bleadperl as change \#18688\.

Arthur

**Comment by p5pRT** [comment]
arthur@contiller.se - Status changed from 'open' to 'resolved'
